File: Kotlin/20_valid_parentheses.kt
fun isValid(s: String): Boolean {
    val stack = ArrayDeque<Char>()
    for (c in s) {
        when (c) {
            '(', '{', '[' -> stack.addLast(c)
            ')', '}', ']' -> {
                if (stack.isEmpty()) return false
                val top = stack.removeLast()
                if ((c == ')' && top != '(') ||
                    (c == '}' && top != '{') ||
                    (c == ']' && top != '[')) {
                    return false
                }
            }
        }
    }
    return stack.isEmpty()
}

File: Dart/20_valid_parentheses.dart
function isValid(s) {
    const stack = [];
    const map = {')': '(', '}': '{', ']': '['};
    for (let c of s) {
        if (Object.values(map).includes(c)) {
            stack.push(c);
        } else if (map[c]) {
            if (stack.pop() !== map[c]) return false;
        }
    }
    return stack.length === 0;
}
